Otp 27

KornelH opened the pull request:
[Erlang/OTP 27](https://www.erlang.org/news/168) introduces two new language elements, [triple-quoted strings](https://www.erlang.org/docs/27/system/data_types.html#string) (docstrings) and [sigils](https://www.erlang.org/docs/27/system/data_types.html#sigil) (type prefix for string literals).

Two simple example modules to test the syntax highlight with the original and updated TextMate language definition:

```erlang
-module(docstring_examples).

-compile(export_all).

-doc """
     Docstring examples
     """.
f() ->
    """
  Line "1"
  Line "2"
  """ = "Line \"1\"
Line \"2\"",
    X = lists:seq(1,3), % just to check is syntax highlight is still ok

    """
      First line starting with two spaces
    Not escaped: "\t \r \xFF" and """

    """ = "  First line starting with two spaces
Not escaped: \"\\t \\r \\xFF\" and \"\"\"
",
    X = lists:seq(1,3), % just to check is syntax highlight is still ok

    """""
""""
FIXME: previous line is not a closing delimiter because opening-closing delimiter have five double quote characters
""""" = "\"\"\"\"",
    X = lists:seq(1,3), % just to check is syntax highlight is still ok

    ok.

-define(THIS_IS_THE_END, "end"). % just to check is syntax highlight is still ok
```

```erlang
-module(sigil_examples).

-compile(export_all).

-doc """
     Sigil examples
     """.
-spec f() -> ok.
f() ->
     ~(monkey ~2..0b\n),
    ~b(monkey ~2..0b\n),
    ~B(monkey ~2..0b\n), % verbatim
    ~s(monkey ~2..0b\n),
    ~S(monkey ~2..0b\n), % verbatim

     ~{monkey ~2..0b\n},
    ~b{monkey ~2..0b\n},
    ~B{monkey ~2..0b\n}, % verbatim
    ~s{monkey ~2..0b\n},
    ~S{monkey ~2..0b\n}, % verbatim

     ~[monkey ~2..0b\n],
    ~b[monkey ~2..0b\n],
    ~B[monkey ~2..0b\n], % verbatim
    ~s[monkey ~2..0b\n],
    ~S[monkey ~2..0b\n], % verbatim

     ~<monkey ~2..0b\n>,
    ~b<monkey ~2..0b\n>,
    ~B<monkey ~2..0b\n>, % verbatim
    ~s<monkey ~2..0b\n>,
    ~S<monkey ~2..0b\n>, % verbatim

     ~/monkey ~2..0b\n/,
    ~b/monkey ~2..0b\n/,
    ~B/monkey ~2..0b\n/, % verbatim
    ~s/monkey ~2..0b\n/,
    ~S/monkey ~2..0b\n/, % verbatim

     ~|monkey ~2..0b\n|,
    ~b|monkey ~2..0b\n|,
    ~B|monkey ~2..0b\n|, % verbatim
    ~s|monkey ~2..0b\n|,
    ~S|monkey ~2..0b\n|, % verbatim

     ~'monkey ~2..0b\n',
    ~b'monkey ~2..0b\n',
    ~B'monkey ~2..0b\n', % verbatim
    ~s'monkey ~2..0b\n',
    ~S'monkey ~2..0b\n', % verbatim

     ~"monkey ~2..0b\n",
    ~b"monkey ~2..0b\n",
    ~B"monkey ~2..0b\n", % verbatim
    ~s"monkey ~2..0b\n",
    ~S"monkey ~2..0b\n", % verbatim

     ~`monkey ~2..0b\n`,
    ~b`monkey ~2..0b\n`,
    ~B`monkey ~2..0b\n`, % verbatim
    ~s`monkey ~2..0b\n`,
    ~S`monkey ~2..0b\n`, % verbatim

     ~#monkey ~2..0b\n#,
    ~b#monkey ~2..0b\n#,
    ~B#monkey ~2..0b\n#, % verbatim
    ~s#monkey ~2..0b\n#,
    ~S#monkey ~2..0b\n#, % verbatim

    X = lists:seq(1,3), % just to check is syntax highlight is still ok
    ok.

-spec g() -> ok. % just to check is syntax highlight is still ok
g() ->
    ~"""
    monkey ~2..0b\n
    business
    """,
    ~b""""
    monkey ~2..0b\n
    """
    business
    """",
    ~B"""""
    monkey ~2..0b\n
    """
    business
    """"",
    ~s""""""
    monkey ~2..0b\n
    """
    business
    """""",
    ~S"""""""
    monkey ~2..0b\n
    """
    business
    """"""",
    X = lists:seq(1,3), % just to check is syntax highlight is still ok

    <<"\"\\µA\""/utf8>> = <<$",$\\,194,181,$A,$">> =
        ~b"""
          "\\µA"
          """ = ~b'"\\µA"' =
        ~B"""
          "\µA"
          """ = ~B<"\µA"> =
        ~"""
          "\µA"
          """ = ~"\"\\µA\"" = ~/"\\µA"/
    X = lists:seq(1,3), % just to check is syntax highlight is still ok

    [$",$\\,$µ,$A,$"] =
        ~s"""
          "\\µA"
          """ = ~s"\"\\µA\"" = ~s["\\µA"] =
        ~S"""
          "\µA"
          """ = ~S("\µA") =
        """
        "\µA"
        """ = "\"\\µA\""
    X = lists:seq(1,3), % just to check is syntax highlight is still ok

    ok.

-define(THIS_IS_THE_END, "end").
```

robertoaloi commented:
Thanks! Not sure if there's a way to have programmatic tests for the grammar. Either way, would it make sense to commit your examples in a dedicated folder, even if only for visual inspection?

robertoaloi commented on Erlang.plist:
```diff
@@ -536,6 +536,219 @@
+				</dict>
+			</dict>
+			<key>comment</key>
+			<string>Only whitespace characters are allowed after the beggining and before the closing sequences and those cannot be in the same line</string>
```
```suggestion
			<string>Only whitespace characters are allowed after the beginning and before the closing sequences and those cannot be in the same line</string>
```

robertoaloi commented:
Before/after, for reference.

Before:

<img width="606" alt="Screenshot 2024-04-02 at 12 24 51" src="https://github.com/erlang-ls/grammar/assets/91769/a3662612-6437-4e44-aa7a-1bbc380f5902">

After:

<img width="612" alt="Screenshot 2024-04-02 at 12 25 58" src="https://github.com/erlang-ls/grammar/assets/91769/101b0bc1-5c5f-4a74-99c4-b52bd856463b">

robertoaloi commented:
@KornelH This is an amazing and very welcome contribution! Let me know when you're ready, so I can merge this. I also like the idea of splitting this into multiple PRs (like you already did for the io:format parts). 🚀

KornelH commented:
Unfortunately, I neither know any automatic way to test syntax highlight. Therefore, I'll add the example files for visual inspection similar to pull request #10. Also, I will break this commit into two separate commits about docstrings and sigils. I don't think these will change in the final version of OTP 27, so I will do it in 1-2 days.

KornelH committed:
Add docstrings as per Erlang/OTP 27

Erlang/OTP 27 [1] introduces triple-quoted strings (docstrings) [2].

[1] https://www.erlang.org/news/168
[2] https://www.erlang.org/docs/27/system/data_types.html#string

KornelH committed:
Add sigils as per Erlang/OTP 27

Erlang/OTP 27 [1] introduces sigils (type prefix for string literals)
[2].

[1] https://www.erlang.org/news/168
[2] https://www.erlang.org/docs/27/system/data_types.html#sigil

garazdawi commented:
As it happens I was also looking into providing the changes in this PR and started looking at how to do CI for the grammar. I've opened #11 with a first draft how we can run tests.

garazdawi commented:
Another things I was looking at was seeing if we can provide an included grammar for docstrings so that they can be highlighted using the Markdown highlighter?

robertoaloi commented:
> Another things I was looking at was seeing if we can provide an included grammar for docstrings so that they can be highlighted using the Markdown highlighter?

That would be awesome!

KornelH commented:
A small note, as far as I understand it, only `-moduledoc` and `-doc` have Markdown as default format, not all triple-quoted strings.
https://erlang.org/documentation/doc-15.0-rc2/doc/system/documentation.html

garazdawi commented:
Any literal string (normal `”`, triple quoted `”””` or sigil `~”””`) in a `doc` or `moduledoc` attribute is by default markdown.

KornelH commented:
Yes, that's ok, any string is by default considered as Markdown in `-doc` and `-moduledoc`, but triple-quoted strings can be used anywhere else. So, what I wanted to say is only that, we shall include Markdown only for these two attributes, not for triple-quoted strings in general.

KornelH commented:
That's true. However in simple strings (`"..."`) all special characters (even `"`) has to be escaped. Embedding Markdown into simple strings (of `-doc` or `-moduledoc`) with escaping all special characters feels a bit complex. In verbatim `-doc` or `-moduledoc` strings (e.g. `~S"..."` or triple-quoted strings) I guess it's doable and useful.

robertoaloi commented:
@KornelH I noticed you opened this for review. Is that intentional or do you still want to break it into separate contributions for sigils and quoted strings? I am happy either way.

KornelH commented:
Well, I can split into two requests.

KornelH commented:
This pull request has been split into pull request #13 and #14.

KornelH commented:
@garazdawi, I was playing a bit with embedding Markdown into `-doc` and `-moduledoc` attributes but the result feels very awkward. Personally, I feel it just disturbs me a lot and totally steals the focus from the source code. The below example is from OTP, `lists:enumerate/1`. Originally, I thought it's a good idea to embed Markdown but now I don't think that anymore. 🙁

![kép](https://github.com/erlang-ls/grammar/assets/9042048/b92b284a-cf82-46e8-b10c-26f300a5688f)

garazdawi commented:
hmm, yeah that was not very nice. Pretty neat that it highlights erlang examples in the markdown though :)

Do you know if it possible to configure the markdown style individually? Would you mind opening a PR (or just pointing to a branch) with what you have so that I can have a look at how it looks?

KornelH commented:
I'm just learning it too and I have no idea if it's possible e.g. set a custom background or shade to embedded languages so it visually would look like a "comment blob" just like eDoc comments, and so it doesn't steal your attention unless you really want to read the documentation. I'll share my grammar changes soon.

garazdawi commented:
It could also be the case this it becomes better if you indent the docs with 4 spaces. That way you get a visual queue that it is something different.

KornelH commented:
@garazdawi, Please find my change in here: https://github.com/KornelH/grammar/commit/309ded82a59e066b3be0218e1f8578c6af9bd2ef

My problem with "indent the docs with 4 spaces" is that, most of the code that I read is not written by me and we cannot enforce the world how to write documentation comments. So, I think most of those will be just like in OTP as shown above. 🫤

garazdawi commented:
> My problem with "indent the docs with 4 spaces" is that, most of the code that I read is not written by me and we cannot enforce the world how to write documentation comments. So, I think most of those will be just like in OTP as shown above. 🫤

Yeah, I was mostly wondering if that fixed the issue. I'll have a look at what you have done and see if I can come up with any clever ideas :)

robertoaloi commented:
I wonder how do other languages (e.g. Elixir) tackle this. Any approach which we can take inspiration from? Would it make sense to use the concept of a "scope" that could then be styled via themes?

robertoaloi commented:
There's also something called [embedded language](https://code.visualstudio.com/api/language-extensions/embedded-languages). Maybe we could leverage that one?

garazdawi commented:
While digging about I found this comment: https://github.com/microsoft/vscode/pull/169956#issuecomment-1370049360

Seems like the way rust/js/ts does it is by having Markdown highlighting as a separate VS Code extension. That way the user can decide if they want it or not and possibly configure how it should look as well.

> There's also something called [embedded language](https://code.visualstudio.com/api/language-extensions/embedded-languages). Maybe we could leverage that one?

This is what @KornelH has done in the prototype.

KornelH commented:
Hi @garazdawi and @robertoaloi,

I reopen this PR just to continue the conversation here. I definitely do not want to ever merge this PR as it was already split into two other PRs and merged before.

A few days ago, I looked into some Elixir sources and saw that, Elixir uses comment look for module and function doc strings. However, they mark the `@moduledoc` and `@doc` keywords as comments too.
![Screenshot 2025-04-22 133917](https://github.com/user-attachments/assets/a14ebdc8-71f8-44a9-966d-7904c3d7700d)

I feel, when docstrings in `-moduledoc` and `-doc` are highlighted as strings, those somewhat steel the focus from the real source code, as those feel to be source code too.
![Screenshot 2025-04-22 133716](https://github.com/user-attachments/assets/69482e96-927e-43b8-bc37-dcb120f53bf6)

I prefer to distinguish source code from documentation comments, as the old eDoc comments did. So, I added a new syntax element to cover `-moduledoc` and `-doc` attributes and mark the content as comment.
![Screenshot 2025-04-22 133603](https://github.com/user-attachments/assets/a23764bf-17ea-4d20-b8af-1a82353c9517)

What do you think?

garazdawi commented:
> What do you think?

I like it.

KornelH commented:
The change is available here: https://github.com/KornelH/grammar/commit/6115175e7e574f69637eaee3db7731915c6f03aa

robertoaloi commented:
@KornelH I see this PR was reopened on April 22. Was that intentional? It looks like this one is superseded by now?

KornelH commented:
Hi @robertoaloi, I reopened it only to continue the conversation. That's all. Let's close this as your new PR continues this topic.
